### DISCUSSION TITLE
Refit with customized reasonphrase

### DISCUSSION THREAD

**Discussion by RHaughton** [reactiveui/refit#1500] [Q&A] [answered]
Hi,

I'm using refit to call a Web API. 
My Web Api doesn't throw exceptions but handles Validation, Not founds, already exists, etc... with Result objects (I'm using the library Fluent Results btw).

My result object has an Error property which is populated whenever an Error occurs. 
Based on this I return a 40X.

How can I customize the ReasonPhrase of the Refit APIResponse to reflect the Error Property of my result object?
In other words how can I customize the ReasonPhrase of Refit ApiResponse to send more meaningfull error message to my frontend?

Thanks for your help

**Comment by glennawatson** [comment, marked as answer]
ReasonPhrase is set by your server, not by Refit. Refit just surfaces whatever came back on the wire.

Two parts:

1. Server side: set the reason phrase on the response. In ASP.NET Core minimal/MVC you can do:

```csharp
httpContext.Features.Get<IHttpResponseFeature>()!.ReasonPhrase = result.Error;
```

Note: ReasonPhrase only exists in HTTP/1.x. Over HTTP/2 and HTTP/3 it is gone, so do not rely on it for transporting error detail. Prefer putting the error in the response body (for example a ProblemDetails or your FluentResults payload) and returning a 4xx status.

2. Client side with Refit: return `ApiResponse<T>` (or catch `ApiException`) and read it:

```csharp
[Post("/things")]
Task<ApiResponse<MyResult>> CreateThing([Body] MyRequest req);

var resp = await api.CreateThing(req);
if (!resp.IsSuccessStatusCode)
{
    var reason = resp.ReasonPhrase;          // HTTP/1.x only
    var body   = resp.Error?.Content;        // the deserialized/raw error body - recommended
}
```

Recommendation: carry the meaningful error in the body and deserialize it, rather than ReasonPhrase, since it is version-dependent and length/charset limited.
